README.md
node-http2
==========

An HTTP/2 ([RFC 7540](http://tools.ietf.org/html/rfc7540))
client and server implementation for node.js.

![Travis CI status](https://travis-ci.org/molnarg/node-http2.svg?branch=master)

**NOTE WELL** This package is officially deprecated. As of node 9.0.0, there is an 'http2' package built-in. You should use that one instead.

Installation
------------

```
npm install http2js
```

API
---

The API is very similar to the [standard node.js HTTPS API](http://nodejs.org/api/https.html). The
goal is the perfect API compatibility, with additional HTTP2 related extensions (like server push).

Detailed API documentation is primarily maintained in the `lib/http.js` file and is [available in
the wiki](https://github.com/molnarg/node-http2/wiki/Public-API) as well.

Examples
--------

### Using as a server ###

```javascript
var options = {
  key: fs.readFileSync('./example/localhost.key'),
  cert: fs.readFileSync('./example/localhost.crt')
};

require('http2js').createServer(options, function(request, response) {
  response.end('Hello world!');
}).listen(8080);
```

### Using as a client ###

```javascript
require('http2js').get('https://localhost:8080/', function(response) {
  response.pipe(process.stdout);
});
```

### Simple static file server ###

An simple static file server serving up content from its own directory is available in the `example`
directory. Running the server:

```bash
$ node ./example/server.js
```

### Simple command line client ###

An example client is also available. Downloading the server's own source code from the server:

```bash
$ node ./example/client.js 'https://localhost:8080/server.js' >/tmp/server.js
```

### Server push ###

For a server push example, see the source code of the example
[server](https://github.com/molnarg/node-http2/blob/master/example/server.js) and
[client](https://github.com/molnarg/node-http2/blob/master/example/client.js).

Status
------

* ALPN is only supported in node.js >= 5.0
* Upgrade mechanism to start HTTP/2 over unencrypted channel is not implemented yet
  (issue [#4](https://github.com/molnarg/node-http2/issues/4))
* Other minor features found in
  [this list](https://github.com/molnarg/node-http2/issues?labels=feature) are not implemented yet

Development
-----------

### Development dependencies ###

There's a few library you will need to have installed to do anything described in the following
sections. After installing/cloning node-http2, run `npm install` in its directory to install
development dependencies.

Used libraries:

* [mocha](http://visionmedia.github.io/mocha/) for tests
* [chai](http://chaijs.com/) for assertions
* [istanbul](https://github.com/gotwarlost/istanbul) for code coverage analysis
* [docco](http://jashkenas.github.io/docco/) for developer documentation
* [bunyan](https://github.com/trentm/node-bunyan) for logging

For pretty printing logs, you will also need a global install of bunyan (`npm install -g bunyan`).

### Developer documentation ###

The developer documentation is generated from the source code using docco and can be viewed online
[here](http://molnarg.github.io/node-http2/doc/). If you'd like to have an offline copy, just run
`npm run-script doc`.

### Running the tests ###

It's easy, just run `npm test`. The tests are written in BDD style, so they are a good starting
point to understand the code.

### Test coverage ###

To generate a code coverage report, run `npm test --coverage` (which runs very slowly, be patient).
Code coverage summary as of version 3.0.1:
```
Statements   : 92.09% ( 1759/1910 )
Branches     : 82.56% ( 696/843 )
Functions    : 91.38% ( 212/232 )
Lines        : 92.17% ( 1753/1902 )
```

There's a hosted version of the detailed (line-by-line) coverage report
[here](http://molnarg.github.io/node-http2/coverage/lcov-report/lib/).

### Logging ###

Logging is turned off by default. You can turn it on by passing a bunyan logger as `log` option when
creating a server or agent.

When using the example server or client, it's very easy to turn logging on: set the `HTTP2_LOG`
environment variable to `fatal`, `error`, `warn`, `info`, `debug` or `trace` (the logging level).
To log every single incoming and outgoing data chunk, use `HTTP2_LOG_DATA=1` besides
`HTTP2_LOG=trace`. Log output goes to the standard error output. If the standard error is redirected
into a file, then the log output is in bunyan's JSON format for easier post-mortem analysis.

Running the example server and client with `info` level logging output:

```bash
$ HTTP2_LOG=info node ./example/server.js
```

```bash
$ HTTP2_LOG=info node ./example/client.js 'https://localhost:8080/server.js' >/dev/null
```

Contributors
------------

The co-maintainer of the project is [Nick Hurley](https://github.com/todesschaf).

Code contributions are always welcome! People who contributed to node-http2 so far:

* [Nick Hurley](https://github.com/todesschaf)
* [Mike Belshe](https://github.com/mbelshe)
* [Yoshihiro Iwanaga](https://github.com/iwanaga)
* [Igor Novikov](https://github.com/vsemogutor)
* [James Willcox](https://github.com/snorp)
* [David Björklund](https://github.com/kesla)
* [Patrick McManus](https://github.com/mcmanus)

Special thanks to Google for financing the development of this module as part of their [Summer of
Code program](https://developers.google.com/open-source/soc/) (project: [HTTP/2 prototype server
implementation](https://google-melange.appspot.com/gsoc/project/details/google/gsoc2013/molnarg/5818821692620800)), and
Nick Hurley of Mozilla, my GSoC mentor, who helped with regular code review and technical advices.

License
-------

The MIT License

Copyright (C) 2013 Gábor Molnár <gabor@molnar.es>


HISTORY.md
Version history
===============

### 3.3.8 (2017-??-??) ###
* Rename package to allow downloading this without deprecation warnings.

### 3.3.7 (2017-09-21) ###
* Mark as incompatible with node >= 9.0.0 (to encourage using the built-in http2 module available by default in node >= 9.0.0).

### 3.3.6 (2016-09-16) ###
* We were not appropriately sending HPACK context updates when receiving SETTINGS_HEADER_TABLE_SIZE. This release fixes that bug.

### 3.3.5 (2016-09-06) ###
* Fix issues with large DATA frames (https://github.com/molnarg/node-http2/issues/207)

### 3.3.4 (2016-04-22) ###
* More PR bugfixes (https://github.com/molnarg/node-http2/issues?q=milestone%3Av3.3.4)

### 3.3.3 (2016-04-21) ###

* Bugfixes from pull requests (https://github.com/molnarg/node-http2/search?q=milestone%3Av3.3.3&type=Issues&utf8=%E2%9C%93)

### 3.3.2 (2016-01-11) ###

* Fix an incompatibility with Firefox (issue 167)

### 3.3.1 (2016-01-11) ###

* Fix some DoS bugs (issues 145, 146, 147, and 148)

### 3.3.0 (2016-01-10) ###

* Bugfix updates from pull requests

### 3.2.0 (2015-02-19) ###

* Update ALPN token to final RFC version (h2).
* Update altsvc implementation to draft 06: [draft-ietf-httpbis-alt-svc-06]

[draft-ietf-httpbis-altsvc-06]: http://tools.ietf.org/html/draft-ietf-httpbis-alt-svc-06

### 3.1.2 (2015-02-17) ###

* Update the example server to have a safe push example.

### 3.1.1 (2015-01-29) ###

* Bugfix release.
* Fixes an issue sending a push promise that is large enough to fill the frame (#93).

### 3.1.0 (2014-12-11) ###

* Upgrade to the latest draft: [draft-ietf-httpbis-http2-16]
 * This involves some state transition changes that are technically incompatible with draft-14. If you need to be assured to interop on -14, continue using 3.0.1

[draft-ietf-httpbis-http2-16]: http://tools.ietf.org/html/draft-ietf-httpbis-http2-16

### 3.0.1 (2014-11-20) ###

* Bugfix release.
* Fixed #81 and #87
* Fixed a bug in flow control (without GitHub issue)

### 3.0.0 (2014-08-25) ###

* Re-join node-http2 and node-http2-protocol into one repository
* API Changes
 * The default versions of createServer, request, and get now enforce TLS-only
 * The raw versions of createServer, request, and get are now under http2.raw instead of http2
 * What was previously in the http2-protocol repository/module is now available under http2.protocol from this repo/module
 * http2-protocol.ImplementedVersion is now http2.protocol.VERSION (the ALPN token)

### 2.7.1 (2014-08-01) ###

* Require protocol 0.14.1 (bugfix release)

### 2.7.0 (2014-07-31) ###

* Upgrade to the latest draft: [draft-ietf-httpbis-http2-14]

[draft-ietf-httpbis-http2-14]: http://tools.ietf.org/html/draft-ietf-httpbis-http2-14

### 2.6.0 (2014-06-18) ###

* Upgrade to the latest draft: [draft-ietf-httpbis-http2-13]

[draft-ietf-httpbis-http2-13]: http://tools.ietf.org/html/draft-ietf-httpbis-http2-13

### 2.5.3 (2014-06-15) ###

* Exposing API to send ALTSVC frames

### 2.5.2 (2014-05-25) ###

* Fix a bug that occurs when the ALPN negotiation is unsuccessful

### 2.5.1 (2014-05-25) ###

* Support for node 0.11.x
* New cipher suite priority list with comformant ciphers on the top (only available in node >=0.11.x)

### 2.5.0 (2014-04-24) ###

* Upgrade to the latest draft: [draft-ietf-httpbis-http2-12]

[draft-ietf-httpbis-http2-12]: http://tools.ietf.org/html/draft-ietf-httpbis-http2-12

### 2.4.0 (2014-04-16) ###

* Upgrade to the latest draft: [draft-ietf-httpbis-http2-11]

[draft-ietf-httpbis-http2-11]: http://tools.ietf.org/html/draft-ietf-httpbis-http2-11

### 2.3.0 (2014-03-12) ###

* Upgrade to the latest draft: [draft-ietf-httpbis-http2-10]

[draft-ietf-httpbis-http2-10]: http://tools.ietf.org/html/draft-ietf-httpbis-http2-10

### 2.2.0 (2013-12-25) ###

* Upgrade to the latest draft: [draft-ietf-httpbis-http2-09]
* [Tarball](https://github.com/molnarg/node-http2/archive/node-http2-2.2.0.tar.gz)

[draft-ietf-httpbis-http2-09]: http://tools.ietf.org/html/draft-ietf-httpbis-http2-09

### 2.1.1 (2013-12-21) ###

* Minor bugfix
* [Tarball](https://github.com/molnarg/node-http2/archive/node-http2-2.1.1.tar.gz)

### 2.1.0 (2013-11-10) ###

* Upgrade to the latest draft: [draft-ietf-httpbis-http2-07][draft-07]
* [Tarball](https://github.com/molnarg/node-http2/archive/node-http2-2.1.0.tar.gz)

[draft-07]: http://tools.ietf.org/html/draft-ietf-httpbis-http2-07

### 2.0.0 (2013-11-09) ###

* Splitting out everything that is not related to negotiating HTTP2 or the node-like HTTP API.
  These live in separate module from now on:
  [http2-protocol](https://github.com/molnarg/node-http2-protocol).
* The only backwards incompatible change: the `Endpoint` class is not exported anymore. Use the
  http2-protocol module if you want to use this low level interface.
* [Tarball](https://github.com/molnarg/node-http2/archive/node-http2-2.0.0.tar.gz)

### 1.0.1 (2013-10-14) ###

* Support for ALPN if node supports it (currently needs a custom build)
* Fix for a few small issues
* [Tarball](https://github.com/molnarg/node-http2/archive/node-http2-1.0.1.tar.gz)

### 1.0.0 (2013-09-23) ###

* Exporting Endpoint class
* Support for 'filters' in Endpoint
* The last time-based release
* [Tarball](https://github.com/molnarg/node-http2/archive/node-http2-1.0.0.tar.gz)

### 0.4.1 (2013-09-15) ###

* Major performance improvements
* Minor improvements to error handling
* [Blog post](http://gabor.molnar.es/blog/2013/09/15/gsoc-week-number-13/)
* [Tarball](https://github.com/molnarg/node-http2/archive/node-http2-0.4.1.tar.gz)

### 0.4.0 (2013-09-09) ###

* Upgrade to the latest draft: [draft-ietf-httpbis-http2-06][draft-06]
* Support for HTTP trailers
* Support for TLS SNI (Server Name Indication)
* Improved stream scheduling algorithm
* [Blog post](http://gabor.molnar.es/blog/2013/09/09/gsoc-week-number-12/)
* [Tarball](https://github.com/molnarg/node-http2/archive/node-http2-0.4.0.tar.gz)

[draft-06]: http://tools.ietf.org/html/draft-ietf-httpbis-http2-06


example/client.js
var fs = require('fs');
var path = require('path');
var http2js = require('..');
var urlParse = require('url').parse;

// Setting the global logger (optional)
http2js.globalAgent = new http2js.Agent({
  rejectUnauthorized: true,
  log: require('../test/util').createLogger('client')
});

// Sending the request
var url = process.argv.pop();
var options = urlParse(url);

// Optionally verify self-signed certificates.
if (options.hostname == 'localhost') {
  options.key = fs.readFileSync(path.join(__dirname, '/localhost.key'));
  options.ca = fs.readFileSync(path.join(__dirname, '/localhost.crt'));
}

var request = process.env.HTTP2_PLAIN ? http2js.raw.get(options) : http2js.get(options);

// Receiving the response
request.on('response', function(response) {
  response.pipe(process.stdout);
  response.on('end', finish);
});

// Receiving push streams
request.on('push', function(pushRequest) {
  var filename = path.join(__dirname, '/push-' + push_count);
  push_count += 1;
  console.error('Receiving pushed resource: ' + pushRequest.url + ' -> ' + filename);
  pushRequest.on('response', function(pushResponse) {
    pushResponse.pipe(fs.createWriteStream(filename)).on('finish', finish);
  });
});

// Quitting after both the response and the associated pushed resources have arrived
var push_count = 0;
var finished = 0;
function finish() {
  finished += 1;
  if (finished === (1 + push_count)) {
    process.exit();
  }
}


example/server.js
var fs = require('fs');
var path = require('path');
var http2js = require('..');

// We cache one file to be able to do simple performance tests without waiting for the disk
var cachedFile = fs.readFileSync(path.join(__dirname, './server.js'));
var cachedUrl = '/server.js';

// The callback to handle requests
function onRequest(request, response) {
  var filename = path.join(__dirname, request.url);

  // Serving server.js from cache. Useful for microbenchmarks.
  if (request.url === cachedUrl) {
    if (response.push) {
      // Also push down the client js, since it's possible if the requester wants
      // one, they want both.
      var push = response.push('/client.js');
      push.writeHead(200);
      fs.createReadStream(path.join(__dirname, '/client.js')).pipe(push);
    }
    response.end(cachedFile);
  }

  // Reading file from disk if it exists and is safe.
  else if ((filename.indexOf(__dirname) === 0) && fs.existsSync(filename) && fs.statSync(filename).isFile()) {
    response.writeHead(200);
    var fileStream = fs.createReadStream(filename);
    fileStream.pipe(response);
    fileStream.on('finish',response.end);
  }

  // Example for testing large (boundary-sized) frames.
  else if (request.url === "/largeframe") {
    response.writeHead(200);
    var body = 'a';
    for (var i = 0; i < 14; i++) {
      body += body;
    }
    body = body + 'a';
    response.end(body);
  }

  // Otherwise responding with 404.
  else {
    response.writeHead(404);
    response.end();
  }
}

// Creating a bunyan logger (optional)
var log = require('../test/util').createLogger('server');

// Creating the server in plain or TLS mode (TLS mode is the default)
var server;
if (process.env.HTTP2_PLAIN) {
  server = http2js.raw.createServer({
    log: log
  }, onRequest);
} else {
  server = http2js.createServer({
    log: log,
    key: fs.readFileSync(path.join(__dirname, '/localhost.key')),
    cert: fs.readFileSync(path.join(__dirname, '/localhost.crt'))
  }, onRequest);
}
server.listen(process.env.HTTP2_PORT || 8080);
